clips: implement direct frame rendering

lordphone opened the pull request:
**Description**

Replace screen capture with direct framebuffer reading. Render to an offscreen `RenderTexture` and read pixels from it using OpenGL instead of capturing a separate UI process. Addresses #36460.

**Details**

1. **Offscreen Rendering**: Creates a `RenderTexture` and uses `begin_texture_mode()` to render the UI to it instead of the default framebuffer.

2. **Pixel Extraction**: Uses OpenGL `glReadPixels` to read RGBA pixels directly from the RenderTexture's framebuffer. Since raylib doesn't provide a function to read from RenderTexture framebuffers, this uses `cffi` to call OpenGL directly.

3. **Format Conversion**: Converts RGBA to RGB24 and flips the image vertically (OpenGL returns bottom-to-top, ffmpeg expects top-to-bottom) using numpy for efficiency.

4. **Streaming**: Pipes raw RGB frames directly to ffmpeg via stdin instead of relying on screen capture.

**Verification**

Demo works: `python tools/clip/run.py --demo`

- Various clip durations
- Quality modes (high/low)
- Title overlay
- Speed multiplier
- Headless mode

^ all manually verified to work

**Notes**

- A GLFW window is created (hidden in headless mode) because GLFW requires a window to create an OpenGL context for RenderTexture operations.

- Prime status fetch errors may appear (`Failed to fetch prime status`) because the temporary environment lacks the `id_rsa` authentication key. These are harmless and don't affect rendering, though on Ubuntu 24.04 the GLFW window may appear frozen when these errors occur (the window is not used for rendering anyway).

- First time contributing! Please let me know if anything needs adjustment.

lordphone committed:
tools: add minimal benchmark script for clip.py performance measurement

lordphone committed:
tools: enhance clip.py with OpenGL integration and improved replay handling

- Added OpenGL support using cffi for frame extraction from render textures.
- Refactored replay readiness check and UI initialization to ensure proper environment setup.
- Updated ffmpeg command to use raw video input from stdin for better performance.
- Improved validation checks for required dependencies and environment variables.
- Cleaned up display management for Xvfb to avoid collisions and ensure stability.

lordphone committed:
tools: optimize frame extraction in clip.py using numpy

- Replaced manual RGBA to RGB conversion with a numpy-based approach for improved performance.
- The new implementation efficiently extracts and flips RGB channels in a single operation.

lordphone committed:
tools: improve OpenGL library loading and enhance replay command

- Explicitly load the OpenGL library on Linux using ctypes for better compatibility.
- Updated the replay command to include the '--headless' option for improved execution in non-GUI environments.

lordphone committed:
tools: refine clip.py for improved process handling and environment setup

- Removed unnecessary PROC_WAIT_SECONDS constant for cleaner code.
- Simplified command extraction logic in check_for_failure function.
- Updated replay command to remove '--headless' option for better compatibility.
- Enhanced display management by using os.environ.pop for cleaner environment variable handling.

lordphone committed:
tools: streamline clip.py by removing warm-up logic and adjusting replay command

- Eliminated SECONDS_TO_WARM constant and associated warm-up logic for a more efficient process.
- Updated replay command to start directly from the specified start time instead of adjusting for warm-up.
- Refined texture management in the GUI initialization for improved performance.

lordphone committed:
tools: remove benchmark script for clip.py performance measurement

lordphone committed:
tools: Fixed static analysis

lordphone commented on tools/clip/run.py:
```diff
@@ -215,56 +263,160 @@ def clip(
+
+    try:
+      # Start replay subprocess
+      with managed_proc(replay_cmd, env) as replay_proc:
```
I see! Thanks for pointing that out.

I kept the replay mostly for the logs, since it publishes real time messages over ZMQ, but I realize in the end I still ended up doing the same thing as before.

I'll try to extract both directly.

lordphone committed:
tools: refactor clip.py to enhance frame handling and mock service integration

- Introduced a MockSubMaster class to simulate service messages for improved testing.
- Removed outdated replay command logic and failure checks for a cleaner implementation.
- Streamlined frame extraction and rendering process, ensuring better handling of missing segments.
- Updated UI initialization and cleanup procedures for enhanced stability and performance.

lordphone committed:
tools:
Clipping with video only works now! enhance clip.py frame handling and resolution determination

- Updated frame resolution retrieval to dynamically use camera segment paths based on quality settings.
- Simplified frame writing to ffmpeg by directly using raw frames, eliminating unnecessary UI rendering steps.
- Added error handling for missing camera segments to ensure robust operation during clipping.

lordphone committed:
tools:
UI overlay showed up! Not finished yet tho.
 enhance clip.py with improved log message handling and frame rendering

- Added functionality to load and sort log messages for better synchronization with frame rendering.
- Implemented processing of log messages during the rendering loop to update UI state dynamically.
- Enhanced frame rendering by converting numpy frames to textures and rendering them directly, improving performance and visual fidelity.
- Introduced error handling for missing camera segments, ensuring robust video clipping operations.

lordphone committed:
Logs not fully working yet. Enhance log message handling in clip.py to support pandaStates

- Added processing for 'pandaStates' in log messages, allowing for dynamic updates to the UI state.
- Implemented a mechanism to copy known fields from old panda states while ensuring compatibility with missing attributes.
- Improved overall robustness of log message handling in the rendering loop.

lordphone committed:
selfdriveState finally updates based by utilyzing selfdrive.test.process_replay.migration, however, no path visualization still.

Enhance clip.py with simulation mode and improved log message handling

- Introduced a simulation mode to bypass SubMaster frequency checks, preventing crashes during testing.
- Patched the `update_msgs` method to ensure correct message handling in simulation.
- Improved log message processing by indexing messages for efficient access and filtering based on time range.
- Updated UI initialization to utilize AugmentedRoadView for better rendering of overlays and camera frames.
- Enhanced error handling for missing camera segments and ensured robust operation during video clipping.

lordphone committed:
All UI elements work now!!!
however, resolution needs fixing.

\Refactor clip.py for improved UI rendering and debugging

- Updated resolution handling to allow dynamic setting based on environment variables.
- Moved GUI initialization to occur after frame reader initialization for better synchronization.
- Added debug logging for liveCalibration and modelV2 messages to assist in troubleshooting.
- Enhanced content rectangle calculations for UI overlays, ensuring proper rendering dimensions.
- Introduced logging of ModelRenderer internal state for performance monitoring during rendering.

lordphone committed:
Enhance clip.py for improved camera frame handling and UI rendering

- Introduced fixed UI resolution for consistent path visualization, aligning with the replay system's approach.
- Updated frame handling to utilize actual camera dimensions for rendering, ensuring accurate display of camera frames.
- Implemented calibration transformations for augmented road view, enhancing the visual fidelity of the UI overlays.
- Improved content rectangle calculations for UI elements, ensuring proper rendering dimensions during video clipping.

lordphone committed:
Applied the correct calibration + aspect ratio scaling from controld and VisionIPC replay systems, now everything finally lines up.

lordphone committed:
tools: now clipping finally shows the front car indicator correctly! clipping is completed it's transformation.

lordphone committed:
Removed all debugging logs.

lordphone committed:
Refactor clip.py to make it readable

- Introduced a ClipGenerator class to encapsulate clipping logic and improve code organization.
- Updated environment setup for Xvfb to ensure proper GUI rendering in headless mode.
- Enhanced frame extraction and rendering process, ensuring accurate display of camera frames with proper calibration.
- Improved command generation for ffmpeg, streamlining video output handling.
- Added validation for input parameters to ensure robust operation during clipping.

lordphone committed:
Fix Static Analysis Errors

lordphone committed:
Two more static analysis.

lordphone committed:
Linting.

lordphone commented:
Alright! I think I have finally gotten this to work. This is probably one of the hardest things I've ever done related to coding. I really hope I wasnt on the wrong track the whole time : ).

**Overview**
Now instead of the old screen record + replay approach, clipping uses the FrameReader to read frames, LogReader to read logs, and combine the elements to directly render a clip.

**Details and Workaround**

Basically, the new clipping tool workflow is:

   1. Load all log and video data.
   2. Monkey-patch the UI SubMaster to accept historical log data.
   3. For each frame, draw the transformed video and the UI (using AugmentedRoadView) to an off-screen canvas.
   4. Extract the finished image and pipe it to ffmpeg.

I ran into lots of small issues on the way, but through combing the codebase and looking at how the original clipping was implemented, I implemented my own fixes and workarounds.

1. To match the same resolution from before (2160x1080 which I believe is the comma3x resolution?), and the path visualization, I copied the calibration and zooming logic for frames from AugmentedRoadView.
2. The demo log files are perhaps old, I used /selfdrive/test/process_replay/migration.py to migrate them over, since the raw logs didnt have any "selfdriveState" logs, which is needed to tell the UI when to turn green and draw path (openpilot engaged). After using the migration tool, everything works.
3. The UI had to be primed with some carParams and liveCalibration messages, or else it didn't think the car was on.
4. Set the Submaster environment with SIUMLATION = 1 so it can accept logs that are not fed in real time.

Known Issue:
FrameReader component can only process HEVC files, so low quality video (qcamera.ts, MPEG-2 Format) can't be extracted.

adeebshihadeh commented:
We're quite busy around here -- but excited to check this out after COMMA_CON!

lordphone committed:
Added support for MacOS by using a raylib hidden window instead of a xvfb headless display for rendering.

Added OpenGL MacOS specific loading.

Closed Framereader after reading a segment.

lordphone committed:
Refactor camera transformation logic in ClipGenerator

Fixed some indentation inconsistencies.

- Introduced a new method `_get_camera_transform` to encapsulate camera calibration and transformation calculations, improving code organization and readability.
- Updated environment variable management in `_setup_environment` for better handling of simulation settings.
- Streamlined the rendering process by utilizing the new transformation method, ensuring accurate camera frame display during video clipping.

lordphone committed:
Refactor ClipGenerator for improved modularity and clarity. No more crazy long functions.

lordphone committed:
Refactor + simplify.
Got rid of some of the redundant comments, fixed new line to follow standard.

lordphone committed:
spelling mistake : (

raysan5 commented:
@lordphone Note that raylib provides functions to get the image from a `RenderTexture` and format it, dealing internally with the multi-platform nuances for different OpenGL versions:

```c
RenderTexture target = LoadRenderTexture(width, height);
...
Image image = LoadImageFromTexture(target.texture);  // Pixel Extraction, GPU -> GPU (slow)
// You get raw RGB24 pixel bytes in image.data, also the .width, .height, .format properties that can vary depending on HighDPI framebuffer and OpenGL version
ImageFlipVertical(&image); // Image flip vertical
ImageFormat(&image, PIXELFORMAT_UNCOMPRESSED_R8G8B8);  // Image format conversion: RGB 24 bpp (only if required, depending on .format)
...
UnloadImage(image);
```

There is no need to use OpeenGL directly.

lordphone committed:
Shifted from OpenGL image fetching to raylib.

During cleanup, a seg fault happens due to a race condetion between gui_app's atexit from applications.py, which destroys the underlying OpenGL graphics context, with cameraview.py that calls __del__ to unload the shader. I had to add a finally block to run.py, and then changed cameraview.py's close() to be more idempotent. No more seg faults running run.py after that.

However, without the changes to cameraview.py, clipping still outputs the video, just crashes during cleanup.

lordphone committed:
frame_texture is now reused instead of unloaded for every frame.

lordphone committed:
Unloading teh pyray image after rendering.

lordphone committed:
Unused Imports

lordphone committed:
Update shader type annotation in CameraView. Fixes mypy.

lordphone committed:
Reversed shader assignment in cameraview.py, and added type ignore for the None assignment instead.

adeebshihadeh commented on tools/clip/run.py:
```diff
@@ -84,71 +76,333 @@ def get_meta_text(lr: LogReader, route: Route):
+    road_view.model_renderer.render(road_view._content_rect)
+    road_view._hud_renderer.render(road_view._content_rect)
+    road_view.alert_renderer.render(road_view._content_rect)
+    road_view.driver_state_renderer.render(road_view._content_rect)
```
why is all this being called here? why can't we just call the `render()` on the main layout?

lordphone replied:
I believe render() needs to connect to a live camerad process via VisionIpcClient and receive video frames, then render those frames to the screen.

(https://github.com/commaai/openpilot/blob/master/selfdrive/ui/onroad/cameraview.py#L186-L188)

Since there's no camerad running in my clip tool, a placeholder returns instead, so I manually draw the frame and call the overlay renderers.

adeebshihadeh commented on tools/clip/run.py:
```diff
@@ -84,71 +76,333 @@ def get_meta_text(lr: LogReader, route: Route):
+    params.put("CarParams", car_params.as_builder().to_bytes())
+  logger.debug('persisted CarParams')
+
+def patched_update_msgs(self, cur_time, msgs):
```
we use this in multiple places now, i think we should just put this in SubMaster directly as a separate mode

adeebshihadeh reviewed (commented):
So this is more correct than before, but it's still a bit too complicated. I propose we replace the linked bounty with something even simpler:

We add `RECORD=1` support to `gui_app`.
* it runs the normal app, but then just writes out the already rendered frames to a video until you quit the app
* it works with anything that uses `gui_app`: spinner, text, ui, etc
* this includes a subset of the pieces to make clips work: exporting the frames and setting up the recording infra 

Once `RECORD=1` is in, it should be way simpler to make the faster-than-realtime version of clips that we're looking for. Let me know if this is makes sense or is still unclear.

TheSecurityDev commented:
> So this is more correct than before, but it's still a bit too complicated. I propose we replace the linked bounty with something even simpler:
> 
> We add `RECORD=1` support to `gui_app`.
> 
> * it runs the normal app, but then just writes out the already rendered frames to a video until you quit the app
> * it works with anything that uses `gui_app`: spinner, text, ui, etc
> * this includes a subset of the pieces to make clips work: exporting the frames and setting up the recording infra
> 
> Once `RECORD=1` is in, it should be way simpler to make the faster-than-realtime version of clips that we're looking for. Let me know if this is makes sense or is still unclear.

I've taken a stab at it here: #36729
